include/boost/numeric/ublas/tensor/tensor.hpp

```diff
@@ -7,615 +7,193 @@
-    }
-
-    friend void swap(basic_tensor& lhs, basic_tensor& rhs) {
-        std::swap(lhs.data_   , rhs.data_   );
-        std::swap(lhs.extents_, rhs.extents_);
-        std::swap(lhs.strides_, rhs.strides_);
-    }
-
-
-    /// \brief return an iterator on the first element of the basic_tensor
-    inline
-    const_iterator begin () const {
-        return data_.begin ();
-    }
-
-    /// \brief return an iterator on the first element of the basic_tensor
-    inline
-    const_iterator cbegin () const {
-        return data_.cbegin ();
-    }
-
-    /// \brief return an iterator after the last element of the basic_tensor
-    inline
-    const_iterator end () const {
-        return data_.end();
-    }
-
-    /// \brief return an iterator after the last element of the basic_tensor
-    inline
-    const_iterator cend () const {
-        return data_.cend ();
-    }
-
-    /// \brief Return an iterator on the first element of the basic_tensor
-    inline
-    iterator begin () {
-        return data_.begin();
-    }
-
-    /// \brief Return an iterator at the end of the basic_tensor
-    inline
-    iterator end () {
-        return data_.end();
-    }
-
-    /// \brief Return a const reverse iterator before the first element of the reversed basic_tensor (i.e. end() of normal basic_tensor)
-    inline
-    const_reverse_iterator rbegin () const {
-        return data_.rbegin();
-    }
-
-    /// \brief Return a const reverse iterator before the first element of the reversed basic_tensor (i.e. end() of normal basic_tensor)
-    inline
-    const_reverse_iterator crbegin () const {
-        return data_.crbegin();
-    }
-
-    /// \brief Return a const reverse iterator on the end of the reverse basic_tensor (i.e. first element of the normal basic_tensor)
-    inline
-    const_reverse_iterator rend () const {
-        return data_.rend();
-    }
-
-    /// \brief Return a const reverse iterator on the end of the reverse basic_tensor (i.e. first element of the normal basic_tensor)
-    inline
-    const_reverse_iterator crend () const {
-        return data_.crend();
-    }
-
-    /// \brief Return a const reverse iterator before the first element of the reversed basic_tensor (i.e. end() of normal basic_tensor)
-    inline
-    reverse_iterator rbegin () {
-        return data_.rbegin();
-    }
-
-    /// \brief Return a const reverse iterator on the end of the reverse basic_tensor (i.e. first element of the normal basic_tensor)
-    inline
-    reverse_iterator rend () {
-        return data_.rend();
-    }
-
-protected:
-
-    extents_type extents_;
-    strides_type strides_;
-    array_type data_;
-};
+        using storage_traits_type        = storage_traits<StorageType>;
+        
+    };
+    
+    template<typename LayoutType, typename StorageType>
+    struct tensor_engine<LayoutType, StorageType>{
+        using extents_type 	        = typename layout::extract_strides_t<void,LayoutType>::extents_type;
```

Comment:
the type of a shape should not depend on an layout type

Reply:
same as https://github.com/boostorg/ublas/pull/90#discussion_r457241360